src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.cs

```diff
@@ -0,0 +1,131 @@
+                                                                                        s_localizableDescriptionRead,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        internal static DiagnosticDescriptor PreferStreamWriteAsyncMemoryOverloadsRule = DiagnosticDescriptorHelper.Create(
+                                                                                        RuleId,
+                                                                                        s_localizableTitleWrite,
+                                                                                        s_localizableMessageWrite,
+                                                                                        DiagnosticCategory.Performance,
+                                                                                        RuleLevel.IdeSuggestion,
+                                                                                        s_localizableDescriptionWrite,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(PreferStreamReadAsyncMemoryOverloadsRule, PreferStreamWriteAsyncMemoryOverloadsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterOperationAction(OnCompilationStart, OperationKind.Invocation);
+        }
+
+        private static void OnCompilationStart(OperationAnalysisContext context)
+        {
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream, out INamedTypeSymbol? stream) ||
+                stream == null)
+            {
+                return;
+            }
+
+            // Verify that the required Memory types were shipped in the current .NET version, otherwise we should not suggest the overload
+            if (context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemoryExtensions) == null ||
+                context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemory1) == null ||
+                context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemReadOnlyMemory1) == null)
+            {
+                return;
+            }
+
+            if (context.Operation is IInvocationOperation invocation &&
+                invocation.TargetMethod is IMethodSymbol method &&
+                IsStreamMethod(method, stream) &&
```

Comment:
It seems like the cheapest check here is to first check the method name, and bail out if it is not "ReadAsync" or "WriteAsync". Do you think we can have this check at the beginning of the invocation callback?

```csharp
   var method = ((IInvocationOperation)context.Operation).TargetMethod;

   DiagnosticDescriptor rule;
   if (string.Equals(method.Name, "ReadAsync", StringComparison.Ordinal))
   {
      rule = PreferStreamReadAsyncMemoryOverloadsRule;
   }
   else if (string.Equals(method.Name, "WriteAsync", StringComparison.Ordinal))
   {
      rule = PreferStreamWriteAsyncMemoryOverloadsRule;
   }
   else
   {
      return;
   }

   if (IsStreamMethod(method, stream) && HasUndesiredArguments(method))
   {
      context.ReportDiagnostic(operation.CreateDiagnostic(rule));
   }
}
```

Reply:
Thanks, @mavasani. I noticed you are suggesting creating the diagnostic on the invocation operation itself, not on the syntax. Would you mind explaining why?

Reply:
@carlossanlop My suggestion was to just use an extension method on IOperation that does the same thing, but is less verbose at call site.

Reply:
Ah good to know there is an extension method, thanks.